pkg/service/storage/ucan_test.go

```diff
@@ -321,3 +334,213 @@ func TestServer(t *testing.T) {
+	require.NoError(t, err)
+
+	// assert the size of the allocation matches our expected size.
+	reader, err := receipt.NewReceiptReaderFromTypes[replica.AllocateOk, fdm.FailureModel](
+		replica.AllocateOkType(), fdm.FailureType(), types.Converters...,
+	)
+	require.NoError(t, err)
+	rcptLink, ok := res.Get(rbi.Link())
+	require.True(t, ok)
+	rcpt, err := reader.Read(rcptLink, res.Blocks())
+	require.NoError(t, err)
+	alloc, err := result.Unwrap(result.MapError(rcpt.Out(), failure.FromFailureModel))
+	require.NoError(t, err)
+	require.Equal(t, expectedSize, alloc.Size)
+
+	// Wait for transfer receipt message, we wait at most 10 seconds (context timeout), or fail.
+	select {
+	case <-ctx.Done():
+		t.Fatal(ctx.Err(), "test did not produce transfer receipt in time")
+	case transferOkMsg := <-transferOkChan:
+		require.NotNil(t, transferOkMsg)
+	}
+}
+
+// startTestHTTPServer starts a simple HTTP server with configurable endpoints.
+func startTestHTTPServer(
+	ctx context.Context,
+	t *testing.T,
+	digest multihash.Multihash,
+	serveData []byte,
+	svc Service,
+	addr, sourcePath, sinkPath, uploadServicePath string,
+) (*http.Server, <-chan message.AgentMessage) {
+	agentCh := make(chan message.AgentMessage, 1)
+	mux := http.NewServeMux()
+
+	// Endpoint to serve data.
+	mux.HandleFunc(fmt.Sprintf("/%s", sourcePath), func(w http.ResponseWriter, r *http.Request) {
+		_, _ = w.Write(serveData)
+	})
+	// Endpoint to store data on the replica.
+	mux.HandleFunc(fmt.Sprintf("/%s", sinkPath), func(w http.ResponseWriter, r *http.Request) {
+		require.NoError(t, svc.Blobs().Store().Put(ctx, digest, uint64(len(serveData)), bytes.NewReader(serveData)))
+		_, _ = w.Write(serveData)
+	})
+	// Endpoint to simulate the upload service.
+	mux.HandleFunc(fmt.Sprintf("/%s", uploadServicePath), func(w http.ResponseWriter, r *http.Request) {
+		roots, blocks, err := car.Decode(r.Body)
+		require.NoError(t, err)
+		bstore, err := blockstore.NewBlockReader(blockstore.WithBlocksIterator(blocks))
+		require.NoError(t, err)
+		agentMessage, err := message.NewMessage(roots, bstore)
+		require.NoError(t, err)
+		agentCh <- agentMessage
+	})
+
+	server := &http.Server{
+		Addr:    addr,
+		Handler: mux,
+	}
+
+	go func() {
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			t.Fatalf("HTTP server ListenAndServe failed: %v", err)
+		}
+	}()
+	time.Sleep(50 * time.Millisecond)
+	t.Cleanup(func() {
+		require.NoError(t, server.Close())
+	})
+	return server, agentCh
+}
+
+// FakePresigned is a stub for upload URL presigning.
+type FakePresigned struct {
+	uploadURL url.URL
+}
+
+func (f *FakePresigned) SignUploadURL(ctx context.Context, digest multihash.Multihash, size, ttl uint64) (url.URL, http.Header, error) {
+	return f.uploadURL, nil, nil
+}
+
+func (f *FakePresigned) VerifyUploadURL(ctx context.Context, url url.URL, headers http.Header) (url.URL, http.Header, error) {
+	// TODO: implement when needed.
+	panic("implement me")
+}
```

Comment:
injected into the storage service, allowing the upload URL to be configured by the test, pointing at the test HTTP server.